util/auto_skip_compressor.cc

```diff
@@ -0,0 +1,122 @@
+    Slice uncompressed_block_data, std::string* compressed_output,
+    const CompressionOptions& opts) {
+  if (compressed_output->size() >
+      (static_cast<uint64_t>(opts.max_compressed_bytes_per_kb) *
+       uncompressed_block_data.size()) >>
+      10) {
+    rejected_count_++;
+  } else {
+    compressed_count_++;
+  }
+  attempted_compression_count_++;
+  if (attempted_compression_count_ >= kWindowSize) {
+    pred_rejection_percentage_ = static_cast<int>(
+        rejected_count_ * 100 / (compressed_count_ + rejected_count_));
+    // fprintf(stdout,
+    //         "[CompressionRejectionProbabilityPredictor::Record] changed "
+    //         "pred_rejection_percentage_: %d\n",
+    //         pred_rejection_percentage_);
+    attempted_compression_count_ = 0;
+    compressed_count_ = 0;
+    rejected_count_ = 0;
+  }
+  return true;
+}
+AutoSkipCompressorWrapper::AutoSkipCompressorWrapper(
+    std::unique_ptr<Compressor> compressor, const CompressionOptions& opts)
+    : CompressorWrapper::CompressorWrapper(std::move(compressor)),
+      exploration_percentage_(10),
+      opts_(opts),
+      rnd_(331),
+      predictor_(
+          std::make_shared<CompressionRejectionProbabilityPredictor>(100)) {}
+
+Status AutoSkipCompressorWrapper::CompressBlock(
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, ManagedWorkingArea* wa) {
+  const int kProbabilityCutOff = 50;
+  bool exploration = rnd_.PercentTrue(exploration_percentage_);
+  // auto setExploration = [&exploration]() { exploration = true; };
+  // (void*)&setExploration;
+  TEST_SYNC_POINT_CALLBACK(
+      "AutoSkipCompressorWrapper::CompressBlock::exploitOrExplore",
+      &exploration);
+  if (exploration) {
+    // fprintf(
+    //     stdout,
+    //     "[AutoSkipCompressorWrapper::CompressBlock] selected:
+    //     exploration\n");
+    Status status = wrapped_->CompressBlock(
+        uncompressed_data, compressed_output, out_compression_type, wa);
+    // determine if it was rejected or compressed
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      predictor_->Record(uncompressed_data, compressed_output, opts_);
+    }
+    return status;
+  } else {
+    auto prediction = predictor_->Predict();
+    // fprintf(stdout,
+    //         "[AutoSkipCompressorWrapper::CompressBlock] selected: exploit "
+    //         "pred_rejection: %d\n",
+    //         prediction);
+    if (prediction < kProbabilityCutOff) {
+      // decide to compress
+      Status status = wrapped_->CompressBlock(
+          uncompressed_data, compressed_output, out_compression_type, wa);
+      // determine if it was rejected or compressed
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        predictor_->Record(uncompressed_data, compressed_output, opts_);
+      }
+      return status;
+    } else {
+      // bypassed compression
+      *out_compression_type = kNoCompression;
+      return Status::OK();
+    }
+  }
```

Comment:
Just to follow up on the conversation on WorkingArea with @pdillinger: I believe the statistics used in this chunk of logic can be moved to WorkingArea if implements the ObtainWorkingArea() to return an implementation of public Compressor::WorkingArea belonging to your AutoSkipCompression. A similar example is `class CompressionContext : public Compressor::WorkingArea` and how it was used. 

The working area will be one per thread for your AutoSkipCompressor therefore your data collection and prediction will only limit to that thread. That also means that you no longer run into awkwardness like using Random variable unsafely. cc @pdillinger

Reply:
For this PR, we decided to make predictor (so auto skip compression manager) not thread safe and amend that in a different PR.